File: src/main/java/com/uml/generator/generator/ComponentDiagramGenerator.java
package com.uml.generator.generator;

import com.uml.generator.analyzer.TraceAggregator;
import com.uml.generator.model.Trace;
import com.uml.generator.renderer.XmiWriter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Generates UML Component Diagrams in XMI 2.5.1 format from aggregated Jaeger
 * traces.
 * Shows services as components with their interfaces and dependencies.
 */
public class ComponentDiagramGenerator implements DiagramGenerator {

    private static final Logger logger = LoggerFactory.getLogger(ComponentDiagramGenerator.class);
    private static final String UML_NAMESPACE = "http://www.eclipse.org/uml2/5.0.0/UML";

    @Override
    public String getDiagramType() {
        return "component";
    }

    @Override
    public String generateXmi(List<Trace> traces) {
        if (traces == null || traces.isEmpty()) {
            logger.warn("No traces provided for component diagram generation");
            return "";
        }

        try {
            TraceAggregator aggregator = new TraceAggregator(traces);
            XmiWriter writer = new XmiWriter();

            // Determine model name
            String modelName = "ComponentDiagram";
            if (!traces.isEmpty() && traces.get(0).getSourceName() != null) {
                modelName = traces.get(0).getSourceName() + "_Component";
            }

            // Create XMI document
            Document doc = writer.createXmiDocument(modelName);
            Element model = writer.getModelElement(doc);

            // Get all services and their relationships
            Set<String> services = aggregator.getAllServices();
            Map<String, Map<String, Object>> serviceMetadata = aggregator.getServiceMetadata();
            Map<String, Map<String, Set<String>>> serviceCalls = aggregator.getServiceCalls();

            // Track created elements to reference them
            Map<String, String> componentIds = new HashMap<>();
            Map<String, String> interfaceIds = new HashMap<>();

            // Create components for all services
            for (String service : services) {
                Map<String, Object> metadata = serviceMetadata.get(service);
                String componentId = createComponent(doc, model, service, metadata);
                componentIds.put(service, componentId);
            }

            // Create interfaces and dependencies
            int interfaceCounter = 0;
            int totalDependencies = 0;

            for (Map.Entry<String, Map<String, Set<String>>> callerEntry : serviceCalls.entrySet()) {
                String callerService = callerEntry.getKey();
                Map<String, Set<String>> calleeMap = callerEntry.getValue();

                for (Map.Entry<String, Set<String>> calleeEntry : calleeMap.entrySet()) {
                    String calleeService = calleeEntry.getKey();
                    Set<String> operations = calleeEntry.getValue();

                    // Create interface for callee service
                    String interfaceName = calleeService + "API";
                    String interfaceId;

                    // Check if interface already exists
                    if (interfaceIds.containsKey(interfaceName)) {
                        interfaceId = interfaceIds.get(interfaceName);
                    } else {
                        interfaceId = createInterface(doc, model, interfaceName, operations);
                        interfaceIds.put(interfaceName, interfaceId);
                    }

                    // Create dependency from caller to interface
                    String callerId = componentIds.get(callerService);
                    if (callerId != null && interfaceId != null) {
                        createDependency(doc, model, callerService + "_uses_" + interfaceName,
                                callerId, interfaceId);
                        totalDependencies++;
                    }

                    // Create interface realization from callee to interface
                    String calleeId = componentIds.get(calleeService);
                    if (calleeId != null && interfaceId != null) {
                        createInterfaceRealization(doc, model, calleeService + "_provides_" + interfaceName,
                                calleeId, interfaceId);
                    }
                }
            }

            logger.info("Generated component diagram XMI with {} service(s) and {} dependencies",
                    services.size(), totalDependencies);

            // Convert to string
            return writer.documentToString(doc);

        } catch (Exception e) {
            logger.error("Failed to generate component diagram XMI", e);
            return "";
        }
    }

    /**
     * Creates a UML Component element.
     */
    private String createComponent(Document doc, Element model, String serviceName, Map<String, Object> metadata) {
        Element component = doc.createElementNS(UML_NAMESPACE, "packagedElement");
        component.setAttribute("xmi:type", "uml:Component");
        String componentId = XmiWriter.generateUUID();
        component.setAttribute("xmi:id", componentId);
        component.setAttribute("name", serviceName);

        // Add stereotype based on service type
        String stereotype = detectServiceStereotype(serviceName, metadata);
        if (stereotype != null && !stereotype.isEmpty()) {
            // Store stereotype as a tagged value comment for now
            // Full stereotype support would require UML Profile extension
            component.setAttribute("visibility", "public");
            // Add comment to indicate stereotype
            addComment(doc, component, "Stereotype: " + stereotype);
        }

        model.appendChild(component);
        return componentId;
    }

    /**
     * Creates a UML Interface element with operations.
     */
    private String createInterface(Document doc, Element model, String interfaceName, Set<String> operations) {
        Element interfaceElement = doc.createElementNS(UML_NAMESPACE, "packagedElement");
        interfaceElement.setAttribute("xmi:type", "uml:Interface");
        String interfaceId = XmiWriter.generateUUID();
        interfaceElement.setAttribute("xmi:id", interfaceId);
        interfaceElement.setAttribute("name", interfaceName);

        // Add operations to interface (limit to 10 for readability)
        if (operations != null && !operations.isEmpty()) {
            int count = 0;
            for (String operation : operations) {
                if (count >= 10)
                    break;
                createOperation(doc, interfaceElement, operation);
                count++;
            }

            // Add comment if there are more operations
            if (operations.size() > 10) {
                addComment(doc, interfaceElement,
                        "+" + (operations.size() - 10) + " more operations not shown");
            }
        }

        model.appendChild(interfaceElement);
        return interfaceId;
    }

    /**
     * Creates a UML Operation element.
     */
    private void createOperation(Document doc, Element interfaceElement, String operationName) {
        Element operation = doc.createElementNS(UML_NAMESPACE, "ownedOperation");
        operation.setAttribute("xmi:id", XmiWriter.generateUUID());

        // Clean operation name
        String cleanName = com.uml.generator.util.NameUtils.cleanOperationName(operationName);
        operation.setAttribute("name", cleanName);
        operation.setAttribute("visibility", "public");

        interfaceElement.appendChild(operation);
    }

    /**
     * Creates a UML Dependency element.
     */
    private void createDependency(Document doc, Element model, String name, String clientId, String supplierId) {
        Element dependency = doc.createElementNS(UML_NAMESPACE, "packagedElement");
        dependency.setAttribute("xmi:type", "uml:Dependency");
        dependency.setAttribute("xmi:id", XmiWriter.generateUUID());
        dependency.setAttribute("name", name);
        dependency.setAttribute("client", clientId);
        dependency.setAttribute("supplier", supplierId);

        model.appendChild(dependency);
    }

    /**
     * Creates a UML InterfaceRealization element.
     */
    private void createInterfaceRealization(Document doc, Element model, String name,
            String implementingClassifierId, String contractId) {
        Element realization = doc.createElementNS(UML_NAMESPACE, "packagedElement");
        realization.setAttribute("xmi:type", "uml:InterfaceRealization");
        realization.setAttribute("xmi:id", XmiWriter.generateUUID());
        realization.setAttribute("name", name);
        realization.setAttribute("implementingClassifier", implementingClassifierId);
        realization.setAttribute("contract", contractId);

        model.appendChild(realization);
    }

    /**
     * Adds a comment element to a UML element.
     */
    private void addComment(Document doc, Element element, String commentText) {
        Element comment = doc.createElementNS(UML_NAMESPACE, "ownedComment");
        comment.setAttribute("xmi:id", XmiWriter.generateUUID());
        comment.setAttribute("annotatedElement", element.getAttribute("xmi:id"));

        Element body = doc.createElementNS(UML_NAMESPACE, "body");
        body.setTextContent(commentText);
        comment.appendChild(body);

        element.appendChild(comment);
    }

    /**
     * Detects the stereotype for a service based on its name and metadata.
     */
    private String detectServiceStereotype(String serviceName, Map<String, Object> metadata) {
        if (serviceName == null) {
            return null;
        }

        String lowerName = serviceName.toLowerCase();

        // Check for frontend/UI services
        if (lowerName.contains("frontend") || lowerName.contains("ui") ||
                lowerName.contains("web") || lowerName.contains("client")) {
            return "WebUI";
        }

        // Check for database services
        if (lowerName.contains("database") || lowerName.contains("db") ||
                lowerName.contains("mongo") || lowerName.contains("postgres") ||
                lowerName.contains("mysql") || lowerName.contains("redis")) {
            return "Database";
        }

        // Check for cache services
        if (lowerName.contains("cache") || lowerName.contains("redis") ||
                lowerName.contains("memcache")) {
            return "Cache";
        }

        // Check metadata for gRPC
        if (metadata != null) {
            Object rpcSystem = metadata.get("rpc.system");
            if (rpcSystem != null && rpcSystem.toString().toLowerCase().contains("grpc")) {
                return "gRPC";
            }
        }

        // Generic microservice for services ending in 'service'
        if (lowerName.endsWith("service")) {
            return "Microservice";
        }

        return null;
    }
}

File: python_version/jaeger_uml_generator/generators/component_diagram_generator.py
"""Component diagram generator with MagicDraw-style output."""

import logging
import xml.etree.ElementTree as ET
from typing import List, Dict, Set
from .diagram_generator import DiagramGenerator
from ..models import Trace
from ..analyzer import TraceAggregator
from ..renderer import XmiWriter, XmiFormat
from ..utils import extract_simple_operation_name


logger = logging.getLogger(__name__)


class ComponentDiagramGenerator(DiagramGenerator):
    """Generates UML Component Diagrams in XMI 2.5.1 format from aggregated Jaeger traces.
    
    This generator produces MagicDraw-style component diagrams with:
    - Package organization for components (Services, Database, etc.)
    - Proper Interface elements with operations
    - Usage dependencies between components and interfaces
    - InterfaceRealization for provided interfaces
    """
    
    # Service categories for package organization
    SERVICE_CATEGORIES = {
        'frontend': 'Presentation',
        'ui': 'Presentation',
        'web': 'Presentation',
        'client': 'Presentation',
        'gateway': 'Presentation',
        'database': 'DataLayer',
        'db': 'DataLayer',
        'mongo': 'DataLayer',
        'postgres': 'DataLayer',
        'mysql': 'DataLayer',
        'redis': 'DataLayer',
        'cache': 'DataLayer',
        'memcache': 'DataLayer',
    }
    
    def __init__(self, xmi_format: str = "papyrus"):
        """Initialize generator with XMI format.
        
        Args:
            xmi_format: Output format ('papyrus' or 'magicdraw')
        """
        format_enum = XmiFormat(xmi_format)
        self.xmi_writer = XmiWriter(format_enum)
    
    def get_diagram_type(self) -> str:
        return "component"
    
    def generate_xmi(self, traces: List[Trace]) -> str:
        """
        Generate XMI for component diagram from multiple traces.
        
        Args:
            traces: List of Trace objects
            
        Returns:
            XMI content as string
        """
        result = self.generate_xmi_with_ids(traces)
        return result.get('xmi_content', '')
    
    def generate_xmi_with_ids(self, traces: List[Trace]) -> Dict[str, any]:
        """
        Generate XMI for component diagram and return element IDs for cross-referencing.
        
        Args:
            traces: List of Trace objects
            
        Returns:
            Dictionary with 'xmi_content', 'component_ids', 'operation_ids'
        """
        if not traces:
            logger.warning("No traces provided for component diagram generation")
            return {'xmi_content': '', 'component_ids': {}, 'operation_ids': {}}
        
        try:
            aggregator = TraceAggregator(traces)
            
            # Determine model name
            model_name = "ComponentDiagram"
            if traces and traces[0].source_name:
                model_name = f"{traces[0].source_name}_Component"
            
            # Create XMI document
            root = self.xmi_writer.create_xmi_document(model_name)
            model = self.xmi_writer.get_model_element(root)
            
            # Get all services and their relationships
            services = aggregator.get_all_services()
            service_metadata = aggregator.get_service_metadata()
            service_calls = aggregator.get_service_calls()
            service_operations = aggregator.get_service_operations()
            
            # Organize services into packages by category
            categorized_services = self._categorize_services(services)
            
            # Track created elements for cross-referencing
            component_ids: Dict[str, str] = {}
            operation_ids: Dict[str, Dict[str, str]] = {}  # service -> {operation -> id}
            
            # Create packages and components with operations
            for category, category_services in categorized_services.items():
                if category_services:
                    # Create package for this category
                    package_elem, package_id = self.xmi_writer.create_package(model, category)
                    
                    # Create components inside the package
                    for service in category_services:
                        metadata = service_metadata.get(service, {})
                        
                        # Get operations for this service directly from aggregator
                        service_ops = service_operations.get(service, set())
                        
                        # Create component with operations
                        component_id, ops_ids = self._create_component_with_operations(
                            package_elem, service, metadata, service_ops
                        )
                        component_ids[service] = component_id
                        operation_ids[service] = ops_ids
            
            # Create simple dependencies between components (no interfaces)
            dependencies_package, _ = self.xmi_writer.create_package(model, "Dependencies")
            
            created_deps: Set[str] = set()
            for caller_service, callee_map in service_calls.items():
                for callee_service, operations in callee_map.items():
                    caller_id = component_ids.get(caller_service)
                    callee_id = component_ids.get(callee_service)
                    
                    if caller_id and callee_id:
                        dep_key = f"{caller_service}_to_{callee_service}"
                        if dep_key not in created_deps:
                            # Create simple Dependency
                            self.xmi_writer.create_usage(
                                dependencies_package,
                                dep_key,
                                caller_id,
                                callee_id
                            )
                            created_deps.add(dep_key)
            
            logger.info(f"Generated component diagram XMI with {len(services)} component(s) "
                       f"and {sum(len(ops) for ops in operation_ids.values())} operation(s)")
            
            return {
                'xmi_content': self.xmi_writer.document_to_string(root),
                'component_ids': component_ids,
                'operation_ids': operation_ids,
                'model': model,
                'root': root
            }
            
        except Exception as e:
            logger.error(f"Failed to generate component diagram XMI: {e}")
            import traceback
            traceback.print_exc()
            return {'xmi_content': '', 'component_ids': {}, 'operation_ids': {}}
    
    def _categorize_services(self, services: Set[str]) -> Dict[str, Set[str]]:
        """
        Categorize services into packages based on naming conventions.
        
        Args:
            services: Set of service names
            
        Returns:
            Dictionary mapping category name to set of services
        """
        categories: Dict[str, Set[str]] = {
            'Presentation': set(),
            'Services': set(),
            'DataLayer': set()
        }
        
        for service in services:
            lower_name = service.lower()
            categorized = False
            
            for keyword, category in self.SERVICE_CATEGORIES.items():
                if keyword in lower_name:
                    categories[category].add(service)
                    categorized = True
                    break
            
            if not categorized:
                categories['Services'].add(service)
        
        # Remove empty categories
        return {k: v for k, v in categories.items() if v}
    
    def _create_component_with_operations(self, parent: ET.Element, service_name: str, 
                                          metadata: Dict[str, any], 
                                          operations: Set[str]) -> tuple:
        """
        Create a UML Component element with ownedOperation elements.
        
        Args:
            parent: Parent element (package)
            service_name: Service name
            metadata: Service metadata
            operations: Set of operation names
            
        Returns:
            Tuple of (component_id, operation_ids_dict)
        """
        # Detect stereotype
        stereotype = self._detect_service_stereotype(service_name, metadata)
        
        # Create component
        component, component_id = self.xmi_writer.create_packaged_element(
            parent, "Component", service_name, visibility="public"
        )
        
        # Add stereotype as comment
        if stereotype:
            self.xmi_writer.add_comment(component, f"«{stereotype}»")
        
        # Add operations directly to component
        operation_ids: Dict[str, str] = {}
        ops_list = list(operations)[:20]  # Limit to 20 for readability
        
        for op_name in ops_list:
            clean_op = extract_simple_operation_name(op_name)
            op_elem, op_id = self.xmi_writer.create_owned_element(
                component, "ownedOperation",
                name=clean_op, visibility="public"
            )
            # Set xmi:type for operation
            op_elem.set(f"{{{self.xmi_writer.XMI_NAMESPACE}}}type", "uml:Operation")
            operation_ids[clean_op] = op_id
        
        # Add comment if more operations exist
        if len(operations) > 20:
            self.xmi_writer.add_comment(
                component,
                f"+{len(operations) - 20} more operations"
            )
        
        return component_id, operation_ids
    
    def _get_provided_operations(self, service_name: str, 
                                 service_calls: Dict[str, Dict[str, Set[str]]]) -> Set[str]:
        """
        Get all operations that a service provides (is called by others).
        
        Args:
            service_name: Service name to check
            service_calls: Service calls mapping
            
        Returns:
            Set of operation names that this service provides
        """
        provided_ops = set()
        
        for caller, callees in service_calls.items():
            if service_name in callees:
                provided_ops.update(callees[service_name])
        
        return provided_ops
    
    def _capitalize_service_name(self, name: str) -> str:
        """Convert service name to PascalCase for interface naming."""
        # Remove common suffixes for cleaner names
        clean_name = name.replace('service', '').replace('Service', '')
        clean_name = clean_name.replace('-', ' ').replace('_', ' ')
        
        # Capitalize each word
        parts = clean_name.split()
        if parts:
            return ''.join(word.capitalize() for word in parts)
        
        return name.capitalize()
    
    def _detect_service_stereotype(self, service_name: str, 
                                   metadata: Dict[str, any]) -> str:
        """
        Detect the stereotype for a service based on its name and metadata.
        
        Args:
            service_name: Service name
            metadata: Service metadata
            
        Returns:
            Stereotype name or empty string
        """
        if not service_name:
            return ""
        
        lower_name = service_name.lower()
        
        # Check for frontend/UI services
        if any(keyword in lower_name for keyword in ['frontend', 'ui', 'web', 'client']):
            return "WebUI"
        
        # Check for gateway services
        if any(keyword in lower_name for keyword in ['gateway', 'api-gateway', 'ingress']):
            return "Gateway"
        
        # Check for database services
        if any(keyword in lower_name for keyword in 
               ['database', 'db', 'mongo', 'postgres', 'mysql']):
            return "Database"
        
        # Check for cache services
        if any(keyword in lower_name for keyword in ['cache', 'redis', 'memcache']):
            return "Cache"
        
        # Check metadata for gRPC
        if metadata:
            rpc_system = metadata.get('rpc.system', '')
            if isinstance(rpc_system, str) and 'grpc' in rpc_system.lower():
                return "gRPC Service"
        
        # Generic microservice for services ending in 'service'
        if lower_name.endswith('service'):
            return "Microservice"
        
        return "Component"
